### DISCUSSION TITLE
How to speed up node startup?

### DISCUSSION THREAD

**Discussion by koteisaev** [nodejs/help#4861] [Ideas]
Hello!
May be you already replied this, but did not found such thread via search.
Context type: Server node application (not a cli tool, not an electron app).

My server app takes like whole minute to just start up, and about 50% of this time is loading of all these packages the server uses.
I seen some obscure hacks based on making pre-ready cache of filenames to speed up the way how require works, but I think should be some other less "hacky" solution to this?

May be some stricter rules flag that will allow to check less of possible paths? like `--speedy-load` ...
Or some way to "prepack" the application with packages to some fewer file(s), or some transpiler that will merge files across project where possible and will speed up the startup by reducing time need to scan actually huge tree of packages?

Or may be some binary format for packages can be added to node/npm where npm package becomes single file with pre-ast-ized binary contents of original js files and BSON for original JSON files, with strict declarations of exposed stuff in metadata, and loading of such submodules of such package done by resolving strict set of pre-known references within such package metadata instead of scanning a lot of possible paths, like `module-name.js`, `module-name.json`, than `module-name/index.js`, and all other paths node checks on file loading?
Such packed packages can have `mpm` file extensions?

**Comment by bl-ue** [comment]
[esbuild](https://github.com/evanw/esbuild) can bundle all your dependencies into one file and minify it; that will probably help.

**Reply by koteisaev** [reply]
Thanks for hint.

**Reply by bl-ue** [reply]
It might be possible to compile the resulting JavaScript file into a (binary?) AST file that can be loaded into V8's memory more quickly 🤔

@addaleax @jasnell is that at all realistic?

**Reply by bl-ue** [reply]
#38905 and related snapshot PRs might be useful too.

**Reply by koteisaev** [reply]
> It might be possible to compile the resulting JavaScript file into a (binary?) AST file that can be loaded into V8's memory more quickly 🤔
> 
> @addaleax @jasnell is that at all realistic?

In theory there are few approaches and some are doable already:
a) pre-AST-ize package sources into gzipped pack and register own module loader the same way how some tooling register custom module loaders,
b) compile source to webassemmbly and in this case the package's `dist/index.js` will be a bootstrapper for webassembly `dist/package.wasm`
c) use webpack, es-build, or other such tooling and ensure it will produce single `dist/index.js` file or such
Actually the most annoying of this is the fact that any of these approaches need some wider adoption to be effective.

**Reply by bl-ue** [reply]
@koteisaev sorry for this late reply, but https://github.com/nodejs/node/issues/35711#issuecomment-790527963 is what you're looking for ;)
